### ISSUE TITLE
Animations for traces

### ISSUE THREAD

**Issue by gravitystorm** [openstreetmap/openstreetmap-website#2132]
Traces can now be imported without requiring the external gpx-import daemon. However, the animation in the `GPX::File.picture` method haven't yet been implemented. This issue is to track progress on this topic and make sure we don't forget to fix this!

We are currently waiting on [dark-panda/gd2-ffij#18](https://github.com/dark-panda/gd2-ffij/issues/18).

**Comment by mmd-osm** [comment]
I believe one thing that could be done in the meantime is to render one distinct image for each frame (=10 different images), which later get passed on to gdImageGifAnimBegin, gdImageGifAnimAdd and gdImageGifAnimEnd to eventually produce the animated gif.

https://github.com/openstreetmap/gpx-import/blob/master/src/image.c#L72-L105

**Comment by mmd-osm** [comment]
By the way, you can use `gem "gd2-ffij", :git => 'git@github.com:mmd-osm/gd2-ffij.git', :branch => 'animated_gif'` to play around with the prototype.

I still get segmentation faults when I'm try to put some kind of loop around the image creation, and I have absolutely no idea why. Somehow Ruby does really strange stuff under the hood. The following creates a gif file with 2 images, though. Maybe there's some GC kicking in, or scoping rules destroy parts of the FFI stuff when iterating in a loop?!

```
    def picture(min_lat, min_lon, max_lat, max_lon, _num_points)
      frames = 10
      width = 250
      height = 250
      proj = OSM::Mercator.new(min_lat, min_lon, max_lat, max_lon, width, height)

      image = []

      image[0] = GD2::Image::IndexedColor.new(width, height) 
      black = image[0].palette.allocate(GD2::Color[0, 0, 0])
      white = image[0].palette.allocate(GD2::Color[255, 255, 255])
      grey = image[0].palette.allocate(GD2::Color[187, 187, 187])

      image[0].draw do |pen|
        pen.color = grey
        pen.rectangle(0, 0, width, height, true)
      end

      image[1] = GD2::Image::IndexedColor.new(width, height)
      black = image[1].palette.allocate(GD2::Color[0, 0, 0])
      white = image[1].palette.allocate(GD2::Color[255, 255, 255])
      grey = image[1].palette.allocate(GD2::Color[187, 187, 187])

      image[1].draw do |pen|
        pen.color = white
        pen.rectangle(0, 0, width, height, true)
      end

      image[1].draw do |pen|
        pen.color = black
        pen.anti_aliasing = true
        pen.dont_blend = false

        oldpx = 0.0
        oldpy = 0.0

        first = true

        points do |p|
          px = proj.x(p.longitude)
          py = proj.y(p.latitude)

          pen.thickness=(3)

          pen.line(px, py, oldpx, oldpy) unless first
            first = false
            oldpy = py
            oldpx = px
        end
      end

      res = GD2::AnimatedGif::gif_anim_begin(image[0])
      res << GD2::AnimatedGif::gif_anim_add(image[0], nil, 50)
      res << GD2::AnimatedGif::gif_anim_add(image[1], image[0], 50) 
      res << GD2::AnimatedGif::gif_anim_end()

      res
    end
```

**Comment by pnorman** [comment]
I'm not a significant user of trace uploads, but how important is this feature?

**Comment by gravitystorm** [comment]
> I'm not a significant user of trace uploads, but how important is this feature?

It means that we can make the built-in trace uploads 100% feature-complete as a replacement for the [external gpx-importer](https://github.com/openstreetmap/gpx-import). This would then allow us to drop that codebase,  enable localized emails for trace imports, as well as running multiple importers in parallel when necessary.

Without the animated gif, it becomes a trade-off between features. With the animated gif, the decision is more straightforward. Since the gif has been around for years, I'm sure some people would complain if it was missing!

**Comment by mmd-osm** [comment]
I finally figured it out how to create 10 frames in a loop without crashing the whole server. You can use the following commit as a template and turn it into more Rails like code. I think there's also some small bug w/r/t with points end up in black or grey in each frame : https://github.com/mmd-osm/openstreetmap-website/commit/fe6f3598f0884e7d52120a41529ec8011e1bfe88

If there's no progress on the external gd2-ffij, you can probably clone my fork, put it on openstreetmap Github org, and adjust the Gemspec accordingly.

Let's retire the external gpx-importer for good.

**Comment by tomhughes** [comment]
If you've got working code why not submit it upstream - I thought that it why you were talking to them on the upstream ticket!

I don't really want to have to fork the gem when upstream is active and receptive.

**Comment by mmd-osm** [comment]
Right, I just don't want to get stuck here in case we can't get anything merged upstream in a reasonable timeframe.

**Comment by gravitystorm** [comment]
This is awesome, great work @mmd-osm! I checked out your branch, [made a couple of stylistic changes](https://github.com/gravitystorm/openstreetmap-website/commit/ab6d657992f89d4578ef02047cd6b43ef4d1dac7) and everything works for me. I can't spot any differences between the animations with your code versus the gpx-importer code, which is great.

However, it looks like there's a problem with the gd2 part of the code, since it fails with a single-point GPX file. We use these in the tests, and running the tests gives an error:

```bash
$ bundle exec ruby -Itest test/models/trace_test.rb 
[Coveralls] Set up the SimpleCov formatter.
[Coveralls] Using SimpleCov's 'rails' settings.
Run options: --seed 30363

# Running:

.........GD Warning: one parameter to a memory allocation multiplication is negative or zero, failing operation gracefully
double free or corruption (top)
Aborted (core dumped)
```

You can see the same thing if you try uploading `test/gpx/fixtures/a.gpx` via the web front end, the webserver dies with the same warning. Could you look into this?

**Comment by mmd-osm** [comment]
Do you see the same crash when skipping https://github.com/gravitystorm/openstreetmap-website/blob/ab6d657992f89d4578ef02047cd6b43ef4d1dac7/lib/gpx.rb#L73-L100 in case num_points == 1?

**Comment by tomhughes** [comment]
I certainly do - the crash is on the call to `GD2::AnimatedGif::gif_anim_add` for the first frame it seems.

**Comment by tomhughes** [comment]
Sorry, for the second frame - the first one in the loop.

**Comment by mmd-osm** [comment]
Replacing

```
 res << GD2::AnimatedGif.gif_anim_add(frames[n], frames[n - 1], delay)
```

by 

```
 res << GD2::AnimatedGif.gif_anim_add(frames[n], frames[n - 1], delay) unless frames[n] == frames[n - 1]
```

lets the unit tests pass. For whatever reason gif_anim_add isn't happy with two identical images in a row. This feels a bit like a hack, and I don't know yet what is caused this behavior.

**Comment by mmd-osm** [comment]
It looks like this is triggering a bug inside the libgd2 library, rather than the gd2-ffij wrapper: if two consecutive frames are identical, they try to create a temporary image of size 0x0. This triggers the "GD Warning", and then the trouble starts by leaving that gdImageGifAnimAddCtx method via some  `goto fail_end;`

To cut a long story short, my proposal would be to skip the gif optimization mode in case both frames are identical: https://github.com/mmd-osm/openstreetmap-website/commit/7f138fb014e201e9e6b63ababf57d0f7c53bd86c

This passes the unit tests while still producing proper images.

---

References:

https://github.com/libgd/libgd/blob/553702980ae89c83f2d6e254d62cf82e204956d0/src/gd_gif_out.c#L723-L727
https://github.com/libgd/libgd/blob/553702980ae89c83f2d6e254d62cf82e204956d0/src/gd_gif_out.c#L793-L800

**Comment by mmd-osm** [comment]
I've opened up an upstream issue with libgd, as valgrind memcheck complains loudly about invalid free operations. They are causing heap corruption issues we're seeing here.

As I don't expect to have this fix deployed on every platform in the near future, I would still highly recommend the approach described in my previous post.

**Comment by tomhughes** [comment]
This was resolved by merging #2204.
